text/3703-wrapping-from.md

```diff
@@ -0,0 +1,277 @@
+- Feature Name: `num_wrapping_from`
+- Start Date: 2024-09-13
+- RFC PR: [rust-lang/rfcs#3703](https://github.com/rust-lang/rfcs/pull/3703)
+- Rust Issue: [rust-lang/rust#0000](https://github.com/rust-lang/rust/issues/0000)
+
+# Summary
+[summary]: #summary
+
+Add a `core::num::WrappingFrom<T>` trait for converting between integer types which might truncate.
+
+
+# Motivation
+[motivation]: #motivation
+
+Today you can convert between arbitrary primitive integer types with `as`.
+
+But there's a few ways in which it'd be nice to have something else:
+
+1. You can't use it in generic code.
+2. You can't use it with non-primitive types like `BigInteger`.
+3. `as` can do enough other things -- like the [evil] `i32::max as u32` -- that you might want to
+   more clearly communicate the intent by using a more specific API.
+
+[evil]: https://play.rust-lang.org/?version=stable&mode=debug&edition=2021&gist=bddbeb692d342b8ebfefa1d25be9275d
+
+This PR thus proposes a new `WrappingFrom` trait for these scenarios:
+
+1. You can `where T: WrappingFrom<U>` in your generics.
+2. The crate defining `BigInteger` can `impl WrappingFrom<BigInteger> for u32 { … }` and such.
+3. You can write `u32::wrapping_from(x)` for a conversion if you prefer, and `u32::from(i32::max)` just won't compile.
+
+
+# Guide-level explanation
+[guide-level-explanation]: #guide-level-explanation
+
+We've seen the `From` trait already, which is for *lossless* and *value-preserving* conversions.
+With integers, though, you might sometimes want *truncating* conversions that you can't do via `From`.
+For that, you can use the `num::WrappingFrom` trait instead.
+
+`WrappingFrom` is specific to *numbers*.  You can't use it for things like `String: From<&str>` which we've used a bunch.
+
+It's is implemented for the numeric widening conversions that also exist in `From`.
+So `i128: From<i32>` and `i128: WrappingFrom<i32>` both exist.
+
+But while `i32: From<i128>` intentionally doesn't exist -- as it'd be lossy -- there *is*
+an `i32: WrappingFrom<i128>` implementation which truncates to just give the bottom 32 bits.
+
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+```rust
+// in core::num (and std::num)
+
+/// Performs potentially-lossy conversions in a quantized numeric lattice.
+///
+/// For an `impl WrappingFrom<T> for U`, if a conceptual value *x* is representable
+/// exactly in both `T` and `U`, then `<U as WrappingFrom<T>>::wrapping_from(x)` must be value-preserving.
+///
+/// If `impl WrappingFrom<T> for U` exists, then `impl WrappingFrom<U> for T` should also exist.
+///
+/// For an integer `a: Big` and a smaller integer type `Small`, truncating with `wrapping_from`
+/// commutes with `wrapping_add`, so `Small::wrapping_from(a).wrapping_add(b)` and
+/// `Small::wrapping_from(a.wrapping_add(b))` will give the same result.
```

Comment:
What is "b" here? I'm guessing it is supposed to be something like "for all `b` in `Small`"?